usvm-ts/src/main/kotlin/org/usvm/machine/expr/TsLocalResolver.kt

```diff
@@ -0,0 +1,199 @@
+                }
+            }
+        }
+
+        // Try to get the saved sort for this dflt object field
+        val savedSort = scope.calcOnState {
+            getSortForDfltObjectField(file, local.name)
+        }
+
+        if (savedSort == null) {
+            // No saved sort means this field was never assigned to, which is an error
+            logger.error { "Trying to read unassigned global variable: '$local' in $file" }
+            scope.assert(falseExpr)
+            return null
+        }
+
+        // Use the saved sort to read the field
+        return mkFieldLValue(savedSort, dfltObject, local.name)
+    }
+
+    // If local is an imported variable:
+    val importInfo = file.importInfos.find { it.name == local.name }
+    if (importInfo != null) {
+        return when (val resolutionResult = scene.resolveImportInfo(file, importInfo)) {
+            is SymbolResolutionResult.Success -> {
+                val importedFile = resolutionResult.file
+                val importedDfltObject = scope.calcOnState { getDfltObject(importedFile) }
+
+                // Initialize globals in the imported file if necessary
+                val isImportedFileGlobalsInitialized = scope.calcOnState { isGlobalsInitialized(importedFile) }
+                if (!isImportedFileGlobalsInitialized) {
+                    logger.info { "Globals are not initialized for imported file: $importedFile" }
+                    scope.doWithState {
+                        initializeGlobals(importedFile)
+                    }
+                    return null
+                }
+
+                if (resolutionResult.exportInfo.name == "*") {
+                    logger.warn { "Star import" }
+                    val lValue = mkFieldLValue(addressSort, importedDfltObject, "__self__")
+                    scope.doWithState {
+                        memory.write(lValue, importedDfltObject, guard = trueExpr)
+                    }
+                    return lValue
+                }
+
+                // Try to get the saved sort for this imported dflt object field
+                val symbolNameInImportedFile = resolutionResult.exportInfo.originalName
+                val savedSort = scope.calcOnState {
+                    getSortForDfltObjectField(importedFile, symbolNameInImportedFile)
+                }
+
+                if (savedSort == null) {
+                    // No saved sort means this field was never assigned to, which is an error
+                    logger.error { "Trying to read unassigned imported symbol: '$local' from '$importedFile'" }
+                    scope.assert(falseExpr)
+                    return null
+                }
+
+                mkFieldLValue(savedSort, importedDfltObject, symbolNameInImportedFile)
+            }
+
+            is SymbolResolutionResult.FileNotFound -> {
+                logger.error { "Cannot resolve import for '$local': ${resolutionResult.reason}" }
+                scope.assert(falseExpr)
```

Comment:
You should forget about this `assert(false)` pattern

Reply:
Unresolved imports are not a hard error. Specifically, on the current state of the development. Some files are missing even in the full sdk, and it is yet unclear what to do in such cases. But throwing a hard error and not producing any execution resulting is a blocker, I wont get any info about coverage in such case. Recall that currently we are still collecting the stats about what we support and we don't by running on real projects, so `throw IllegalStateException("sorry, your project or/and IR is garbage, fix it and come again for any results")` is a no-go solution.